src/DI.Abstractions/Extensions/ServiceCollectionDescriptorExtensions.cs

```diff
@@ -684,6 +684,110 @@ public static void TryAddSingleton<TService>(this IServiceCollection collection,
+        public static IServiceCollection Replace(this IServiceCollection services, Type serviceType, Type implementationType)
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            if (serviceType == null)
+            {
+                throw new ArgumentNullException(nameof(serviceType));
+            }
+
+            if (implementationType == null)
+            {
+                throw new ArgumentNullException(nameof(implementationType));
+            }
+
+            if (!services.TryGetDescriptors(serviceType, out var descriptors))
+            {
+                throw new ArgumentException(Resources.FormatNoServiceRegistered(serviceType), nameof(serviceType));
+            }
+
+            foreach (var descriptor in descriptors)
+            {
+                var index = services.IndexOf(descriptor);
+
+                // Copy the old descriptor, but replace its implementation type.
+                var newDescriptor = ServiceDescriptor.Describe(
+                    descriptor.ServiceType,
+                    implementationType,
+                    descriptor.Lifetime);
+
+                services.Insert(index, newDescriptor);
+
+                services.Remove(descriptor);
+            }
+
+            return services;
+        }
+
+        /// <summary>
+        /// Tries to get all registered descriptors with the specified <typeparamref name="TService"/>.
+        /// </summary>
+        /// <typeparam name="TService">The type of the services to get.</typeparam>
+        /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="descriptors">The matching <see cref="ServiceDescriptor"/>s.</param>
+        public static bool TryGetDescriptors<TService>(this IServiceCollection services, out ICollection<ServiceDescriptor> descriptors)
+        {
+            return services.TryGetDescriptors(typeof(TService), out descriptors);
+        }
+
+        /// <summary>
+        /// Tries to get all registered descriptors with the specified <paramref name="serviceType"/>.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="serviceType">The type of the services to get.</param>
+        /// <param name="descriptors">The matching <see cref="ServiceDescriptor"/>s.</param>
+        public static bool TryGetDescriptors(this IServiceCollection services, Type serviceType, out ICollection<ServiceDescriptor> descriptors)
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            if (serviceType == null)
+            {
+                throw new ArgumentNullException(nameof(serviceType));
+            }
+
+            return (descriptors = services.Where(service => service.ServiceType == serviceType).ToList()).Count > 0;
```

Comment:
Split this up into multiple lines. Treating assignment like an expression always looks messy IMO.